lib/livebook/runtime/node_pool.ex

```diff
@@ -0,0 +1,102 @@
+defmodule Livebook.Runtime.NodePool do
+  use GenServer
+
+  @moduledoc false
+
+  # A pool for reusing child node names.
+  #
+  # `own_pool` refers to the list of unused names.
+  # `given_pool` refers to the list of names ever generated.
+  #
+  # `buffer_time` refers to time the pool waits before
+  # adding a name to `pool`,  which by default is 1 minute.
+
+  @default_time 60_000
+
+  # Client interface
+
+  @doc """
+  Starts the GenServer from a Supervision tree
+  """
+  def start_link(opts) do
+    # Note that the following options are for testing only.
+    name = opts[:name] || __MODULE__
+    buffer_time = opts[:buffer_time] || @default_time
+    given_pool = opts[:given_pool] || []
+    own_pool = opts[:own_pool] || []
+
+    GenServer.start_link(
+      __MODULE__,
+      %{buffer_time: buffer_time, given_pool: given_pool, own_pool: own_pool},
```

Comment:
The pools are actually internal state and not options, and that would normally be initialized in the `init` callback. So here you could just pass the public options like `GenServer.start_link(__MODULE__, buffer_time: buffer_time, name: name)`, and then build the initial state in `init`.

I'd suggest more informative names, like `generated_names` and `free_names`.

Reply:
I am aware that pools are internal state and not user provided options. However I used the options to set the states for testing.. I suppose, I will have to take the tedious method of overriding state with ~~sys:replace_state~~.

Reply:
I mean, given the state is internal we shouldn't really care what it is. The test case would just simulate taking the name, returning it and check if the next time we get the same name.

By setting/asserting the state you effectively test the private functions - which theoretically you could, but they are private exactly because all of that is just an implementation detail that could change at any point as long as the semantics are preserved :)